core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -528,22 +533,38 @@ private TaskInfo doUpdateTask(
             }
         }
 
-        fragment.map(PlanFragment::getActiveCatalogs)
-                .ifPresent(activeCatalogs -> {
-                    Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
-                            .map(CatalogProperties::getCatalogHandle)
-                            .collect(toImmutableSet());
-                    if (sqlTask.setCatalogs(catalogHandles)) {
-                        ReentrantReadWriteLock.ReadLock catalogInitLock = catalogsLock.readLock();
-                        catalogInitLock.lock();
-                        try {
-                            connectorServicesProvider.ensureCatalogsLoaded(session, activeCatalogs);
-                        }
-                        finally {
-                            catalogInitLock.unlock();
-                        }
+        Optional<List<CatalogProperties>> activeCatalogsOpt = fragment.map(PlanFragment::getActiveCatalogs);
+        if (activeCatalogsOpt.isPresent()) {
+            List<CatalogProperties> activeCatalogs = activeCatalogsOpt.get();
+            Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
+                    .map(CatalogProperties::getCatalogHandle)
+                    .collect(toImmutableSet());
+            if (sqlTask.setCatalogs(catalogHandles)) {
+                ListenableFuture<Void> catalogLoading = Futures.submit(() -> {
+                    ReentrantReadWriteLock.ReadLock catalogInitLock = catalogsLock.readLock();
+                    catalogInitLock.lock();
+                    try {
+                        connectorServicesProvider.ensureCatalogsLoaded(session, activeCatalogs);
                     }
-                });
+                    finally {
+                        catalogInitLock.unlock();
+                    }
+                }, catalogLoadingExecutor);
+                try {
+                    catalogLoading.get(5, SECONDS);
```

Comment:
this will introduce additional query latency (task won't be updated with splits for 5s as only one request at a time is allowed). We need to figure another way to do this.

Reply:
> My understanding is that this will only wait 5 seconds if the catalogs are still loading ( and the query cannot execute anyway)

Do I understand correctly that catalogs can be different per query? What if catalogs load in 10 milliseconds? In that case split assignments will be blocked for remainder of 5s.

Lastly, is 5s always sufficient?

Reply:
Is that how the future timeout works? It always blocks 5s?

5s is not always sufficient for loading all catalogs, which is why I need to block execution until the task has initialized all catalogs.

Reply:
Just tested. I believe this early terminates

Reply:
> 5s is not always sufficient for loading all catalogs, which is why I need to block execution until the task has initialized all catalogs.

I think this means it should not be time based, but event based (e.g. some future)